### DISCUSSION TITLE
How can I add Hardhat to my NPM Package?

### DISCUSSION THREAD

**Discussion by TechnicallyWeb3** [NomicFoundation/hardhat#6004] [General]
I have written a hardhat project which I would like to deploy as a library.

Within my Hardhat project I can use the wttp.fetch("technicallyweb3.eth") function which relies heavily on hardhat and uses a specific hardhat config to operate. 

My goal is to allow people to use npm install wttp-handler, import the default wttp handler and allow devs to use the fetch function as easily as the built-in fetch function without needing to install and configure hardhat. I've already published to wttp-handler but getting all kinds of errors when tying to execute.

Furthermore, my hardhat configuration shouldn't conflict if the parent project has it's own hardhat installation.

I've never tried using hardhat like this before, and don't even know if it's intended to work like this, but when running my library in a js file I get a hardhat structure error.

Has anyone set up this type of static library using hardhat? Any tips for my first go?

The project I'm trying to convert into a library is https://github.com/technicallyweb3/wttp (npm-library branch) and the package is https://www.npmjs.com/package/wttp-handler

**Comment by TechnicallyWeb3** [comment]
Was intending this to be more a discussion if what I'm trying to do is possible, but since no one asked, here's the output when running some code using my library:

```
C:\Projects\TechnicallyWeb3\example-wttp\node_modules\hardhat\internal\core\project-structure.js:39
        throw new errors_1.HardhatError(errors_list_1.ERRORS.GENERAL.NOT_INSIDE_PROJECT);
        ^

HardhatError: HH1: You are not inside a Hardhat project.
    at getUserConfigPath (C:\Projects\TechnicallyWeb3\example-wttp\node_modules\hardhat\internal\core\project-structure.js:39:15)   
    at resolveConfigPath (C:\Projects\TechnicallyWeb3\example-wttp\node_modules\hardhat\internal\core\config\config-loading.js:37:64)
    at loadConfigAndTasks (C:\Projects\TechnicallyWeb3\example-wttp\node_modules\hardhat\internal\core\config\config-loading.js:54:18)
    at Object.<anonymous> (C:\Projects\TechnicallyWeb3\example-wttp\node_modules\hardhat\internal\lib\hardhat-lib.js:30:84)
    at Module._compile (node:internal/modules/cjs/loader:1469:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12)
    at Module.require (node:internal/modules/cjs/loader:1311:19)
    at require (node:internal/modules/helpers:179:18) {
  parent: undefined,
  _stack: 'HardhatError: HH1: You are not inside a Hardhat project.\n' +
    '    at getUserConfigPath (C:\\Projects\\TechnicallyWeb3\\example-wttp\\node_modules\\hardhat\\internal\\core\\project-structure.js:39:15)\n' +
    '    at resolveConfigPath (C:\\Projects\\TechnicallyWeb3\\example-wttp\\node_modules\\hardhat\\internal\\core\\config\\config-loading.js:37:64)\n' +
    '    at loadConfigAndTasks (C:\\Projects\\TechnicallyWeb3\\example-wttp\\node_modules\\hardhat\\internal\\core\\config\\config-loading.js:54:18)\n' +
    '    at Object.<anonymous> (C:\\Projects\\TechnicallyWeb3\\example-wttp\\node_modules\\hardhat\\internal\\lib\\hardhat-lib.js:30:84)\n' +
    '    at Module._compile (node:internal/modules/cjs/loader:1469:14)\n' +
    '    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)\n' +
    '    at Module.load (node:internal/modules/cjs/loader:1288:32)\n' +
    '    at Module._load (node:internal/modules/cjs/loader:1104:12)\n' +
    '    at Module.require (node:internal/modules/cjs/loader:1311:19)\n' +
    '    at require (node:internal/modules/helpers:179:18)',
  errorDescriptor: {
    number: 1,
    message: 'You are not inside a Hardhat project.',
    title: 'You are not inside a Hardhat project',
    description: 'You are trying to run Hardhat outside of a Hardhat project.\n' +
      '\n' +
      'You can learn how to use Hardhat by reading the [Getting Started guide](/hardhat-runner/docs/getting-started).',
    shouldBeReported: false
  },
  number: 1,
  messageArguments: {},
  _isHardhatError: true
}

Node.js v20.17.0
```

You can get here by installing:
```npm install wttp-handler@2.0.1```
and running this code:
```
const { wttp } = require("wttp-handler");
wttp.fetch("wttp://0x3B370D271344d0C4d5DCbd80Bfe714C0e1ED648a").then((response) => {
  console.log(response);
});
```

**Comment by kanej** [comment]
Hardhat 2 assumes on file project structure around the setup of `hardhat.config.{ts,js}`, meaning it is not going to work easily as a library.
Have you considered connecting directly using a library like `ethers` or `viem` without the setup that Hardhat provides?

**Reply by TechnicallyWeb3** [reply]
I have, and this is what I ended up doing, however hardhat as a solution in my library would have been awesome since I had to basically rebuild a bunch of tools that are native to hardhat or have hardhat extensions just to get it to function properly as a standalone library. 

Maybe on the wish list for hardhat 3? Bundled config so they can call my library with a custom config separate from the hardhat on the project level. Eventually the library will need to deploy contracts, switch networks and interact with. I would love to create ignition scripts which can be used by my library for complex tasks, again keeping a specific configuration for my packages. I love building hardhat scripts to make doing things easily, things I would like to extend to the actual application. 

Anyone else want this or just me?

**Reply by kanej** [reply]
In Hardhat 3 you can create a Hardhat Runtime Environment (hre) based on a config object you pass in programmatically. Take a look at the Alpha: https://hardhat.org/hardhat3-alpha
